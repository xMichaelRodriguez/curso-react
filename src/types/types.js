export const types = {
  //ui
  uiOpenModal: "[UI] Open modal",
  uiCloseModal: "[UI] Close modal",

  //event
  eventSetActive: "[EVENT] Set Active",
  eventAddNew: "[EVENT] Add new",
  eventClearActiveEvent: "[EVENT] Clear active event",
  eventUpdated: "[EVENT] Event updated",
  eventDeleted: "[EVENT] Event deleted",

  //auth

  authFinishChecking: "[AUTH] Finish checking login state",
  authStartLogin: "[AUTH] Start login",
  authLogin: "[AUTH] logged",
  authStartregister: "[AUTH] Start register",
  authStartTokenRenew: "[AUTH] Start token renew",
  authLogout: "[AUTH] Auth logout",
};
