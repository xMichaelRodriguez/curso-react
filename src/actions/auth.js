import Swal from "sweetalert2";
import { fetchAsync, fetchSync } from "../helpers/fetch";
import { types } from "../types/types";
import { eventLogout } from "./events";

export const startLogin = (email, password) => {
  return async (dispatch) => {
    Swal.fire({
      title: "Iniciando Sesion...",
      text: "Por favor espere...",
      allowOutsideClick: false,
      allowEnterKey: false,
      allowEscapeKey: false,
      showConfirmButton: false,
      onBeforeOpen: () => {
        Swal.showLoading();
      },
    });
    const resp = await fetchSync("auth", { email, password }, "POST");

    const body = await resp.json();

    if (body.ok) {
      localStorage.setItem("token", body.token);
      localStorage.setItem("token-initDate", new Date().getTime());
      dispatch(login({ uid: body.uid, name: body.name }));
      Swal.close();
    } else {
      Swal.close();
      Swal.fire("Error", body.msg, "error");
    }
  };
};

export const startRegister = (name, email, password) => {
  return async (dispatch) => {
    Swal.fire({
      title: "Registrando...",
      text: "Por favor espere...",
      allowOutsideClick: false,
      allowEnterKey: false,
      allowEscapeKey: false,
      showConfirmButton: false,
      onBeforeOpen: () => {
        Swal.showLoading();
      },
    });
    const resp = await fetchSync("auth/new", { name, email, password }, "POST");

    const body = await resp.json();

    if (body.ok) {
      localStorage.setItem("token", body.token);
      localStorage.setItem("token-initDate", new Date().getTime());
      dispatch(login({ uid: body.uid, name: body.name }));
      Swal.close();
    } else {
      Swal.close();
      Swal.fire("Error", body.msg, "error");
    }
  };
};

export const startChecking = () => {
  return async (dispatch) => {
    const resp = await fetchAsync("auth/renew");

    const body = await resp.json();

    if (body.ok) {
      localStorage.setItem("token", body.token);

      localStorage.setItem("token-initDate", new Date().getTime());

      dispatch(login({ uid: body.uid, name: body.name }));

      Swal.close();
    } else {
      dispatch(checkingFish());
    }
  };
};

const checkingFish = () => ({
  type: types.authFinishChecking,
});

const login = (user) => ({
  type: types.authLogin,
  payload: user,
});

export const startLogout = () => {
  return (dispatch) => {
    localStorage.clear();

    dispatch(eventLogout());

    dispatch(logout());
  };
};

const logout = () => ({
  type: types.authLogout,
});
