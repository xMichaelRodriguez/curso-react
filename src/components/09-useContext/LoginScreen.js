import React, { useContext } from "react";
import { UserContext } from "./UserContext";

export const LoginScreen = () => {
  //Obtener referencia al context
  const { setUser } = useContext(UserContext);

  const handleLogin = () => {
    setUser({
      id: new Date().getTime(),
      name: "Michael Rodriguez",
      email: "scottlovos503@gmail.com",
    });
  };
  return (
    <div>
      <h1>Login PAGE</h1>
      <hr />
      <button className="btn btn-primary" type="button" onClick={handleLogin}>
        Login
      </button>
    </div>
  );
};
