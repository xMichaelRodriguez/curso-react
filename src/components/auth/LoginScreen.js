import React from "react";
import { FaSignInAlt } from "react-icons/fa";
import { Link } from "react-router-dom";
export const LoginScreen = () => {
  const handleLogin = (e) => e.preventDefault();
  return (
    <div className="shadow auth">
      <form onSubmit={handleLogin}>
        <h1 className="mb-4">
          <ins>Blog</ins>
        </h1>

        <div className="form-row">
          <div className="col-md-12 mb-3">
            <div className="form-group">
              <label>Email</label>
              <input
                className="form-control"
                type="email"
                placeholder="example@dominio.com"
              />
            </div>
          </div>
          <div className="col-md-12 mb-3">
            <label>Password</label>
            <input
              type="password"
              className="form-control"
              placeholder="Password"
            />
          </div>
          <div className="col-md-6 mb-3">
            <button className="btn primary btn-lg" type="submit">
              <FaSignInAlt /> &nbsp;
              <span>Login</span>
            </button>
          </div>
          <div className="col-md-6 w-100">
            <Link to="/auth/register" class="btn btn-link text-capitalize">
              forgot password?
            </Link>
          </div>
        </div>
      </form>
    </div>
  );
};
