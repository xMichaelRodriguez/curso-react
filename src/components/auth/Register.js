import React from 'react';
import { FaSignInAlt } from 'react-icons/fa';
import { Link } from 'react-router-dom';

export const Register = () => {
  const handleRegister = (e) => e.preventDefault();
  return (
    <div className='shadow auth animate__animated animate__fadeIn'>
      <form onSubmit={handleRegister}>
        <h1 className='mb-4'>
          <ins>Blog</ins>
        </h1>

        <div className='form-row'>
          <div className='col-md-6'>
            <div className='form-group'>
              <label>First Name</label>
              <input
                className='form-control'
                type='text'
                placeholder='First Name'
              />
            </div>
          </div>
          <div className='col-md-6'>
            <div className='form-group'>
              <label>Last Name</label>
              <input
                className='form-control'
                type='text'
                placeholder='Last Name'
              />
            </div>
          </div>
          <div className='col-md-12 mb-3'>
            <div className='form-group'>
              <label>Email</label>
              <input
                className='form-control'
                type='email'
                placeholder='example@dominio.com'
              />
            </div>
          </div>
          <div className='col-md-12 mb-3'>
            <label>Password</label>
            <input
              type='password'
              className='form-control'
              placeholder='Password'
            />
          </div>
          <div className='col-md-6 mb-3'>
            <button className='btn primary btn-lg' type='submit'>
              <FaSignInAlt /> &nbsp;
              <span>Register</span>
            </button>
          </div>
          <div className='col-md-6 w-100'>
            <Link to='/auth/' class='btn btn-link text-capitalize'>
              I have an account
            </Link>
          </div>
        </div>
      </form>
    </div>
  );
};
