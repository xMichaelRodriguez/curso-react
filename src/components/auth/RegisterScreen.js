import React from "react";
import { Link } from "react-router-dom";

export const RegisterScreen = () => {
  return (
    <>
      <h3 className="auth__title mb-5">Register</h3>
      <form>
        <div className="auth__group">
          <input
            type="text"
            className="auth__input"
            name="firstName"
            placeholder="Your  First Name"
            autoComplete="false"
          />
          <span class="highlight"></span>
          <span class="bar"></span>
          <label>First Name</label>
        </div>
        <div className="auth__group">
          <input
            type="text"
            className="auth__input"
            name="lastName"
            placeholder="Your Last Name"
            autoComplete="false"
          />
          <span class="highlight"></span>
          <span class="bar"></span>
          <label>Last Name</label>
        </div>
        <div className="auth__group">
          <input
            type="email"
            className="auth__input"
            name="email"
            placeholder="Your Email"
            autoComplete="false"
          />
          <span class="highlight"></span>
          <span class="bar"></span>
          <label>Email</label>
        </div>
        <div className="auth__group">
          <input
            type="Password"
            className="auth__input"
            name="password"
            placeholder="Your Password"
          />
          <span class="highlight"></span>
          <span class="bar"></span>
          <label>Password</label>
        </div>

        <button type="submit" className="btn btn-primary btn-block mb-5">
          Sign In
        </button>

        <Link to="/auth/login" className="link">
          Already registered?
        </Link>
      </form>
    </>
  );
};
