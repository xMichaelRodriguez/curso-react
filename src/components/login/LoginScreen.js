import React, { useContext } from "react";
import { AuthContext } from "../../auth/AuthContext";
import { types } from "../../types/types";

export const LoginScreen = ({ history }) => {
  const { dispatch } = useContext(AuthContext);
  const handleLogin = () => {
    // history.push('/');
    // history.replace('/');
    const lastPath = localStorage.getItem("lastPath") || "/";
    dispatch({
      type: types.login,
      payload: {
        name: "Michael Rodriguez",
      },
    });

    history.replace(lastPath);
  };

  return (
    <div className="container mt-5">
      <h1>Login</h1>
      <hr />

      <button className="btn btn-primary" onClick={handleLogin}>
        Login
      </button>
    </div>
  );
};
