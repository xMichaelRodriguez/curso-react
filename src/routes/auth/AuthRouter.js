import React from "react";
import { Switch, Redirect, Route } from "react-router-dom";
import { LoginScreen } from "../../components/auth/LoginScreen";
import { Register } from "../../components/auth/Register";
export const AuthRouter = () => {
  return (
    <div>
      <Switch>
        <Route exact path="/auth/login" component={LoginScreen} />
        <Route exact path="/auth/register" component={Register} />

        <Redirect  to="/auth/login" />
      </Switch>
    </div>
  );
};
