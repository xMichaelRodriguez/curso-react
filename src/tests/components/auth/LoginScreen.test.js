import React from 'react';
import '@testing-library/jest-dom/extend-expect';
import { fireEvent, render } from '@testing-library/react';
import { LoginScreen } from '../../../components/login/LoginScreen';
import { AuthContext } from '../../../auth/AuthContext';
import { types } from '../../../types/types';
describe('Tests in <LoginScreen/>', () => {

    const contextValue = {
        dispatch: jest.fn(),

    }
    const history = {
        replace: jest.fn()
    }
    let component;
    beforeEach(() => {
        component = render(
            <AuthContext.Provider value={contextValue}>
                <LoginScreen history={history} />
            </AuthContext.Provider>
        )
    })

    test('should display currently', () => {

        const component = render(
            <AuthContext.Provider value={contextValue}>
                <LoginScreen history={history} />
            </AuthContext.Provider>
        )

        const el = component.container
        expect(el).toMatchSnapshot()

    });

    test('must perform dispatch and navigation', () => {

        const el = component.container.querySelector('button');
        fireEvent.click(el);
        expect(contextValue.dispatch).toHaveBeenLastCalledWith({
            type: types.login,
            payload: {
                name: "Michael Rodriguez",
            }
        });

        expect(history.replace).toHaveBeenCalledWith('/')
        localStorage.setItem('lastPath', '/dc');
        fireEvent.click(el);
        expect(history.replace).toHaveBeenCalledWith('/dc')


    });

});
