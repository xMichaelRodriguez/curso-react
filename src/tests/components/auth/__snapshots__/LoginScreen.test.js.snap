// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in <LoginScreen/> should display currently 1`] = `
<div>
  <div
    class="container mt-5"
  >
    <h1>
      Login
    </h1>
    <hr />
    <button
      class="btn btn-primary"
    >
      Login
    </button>
  </div>
</div>
`;
